File: src/main/java/com/fengsheng/phase/ReceivePhaseSenderSkill.java
package com.fengsheng.phase;

import com.fengsheng.Fsm;
import com.fengsheng.Player;
import com.fengsheng.ResolveResult;
import com.fengsheng.card.Card;

import java.util.List;

/**
 * 接收情报时，传出者的技能
 *
 * @param whoseTurn     谁的回合
 * @param messageCard   情报牌
 * @param receiveOrder  接收情报牌的顺序（也就是后续结算死亡的顺序）
 * @param inFrontOfWhom 情报在谁面前
 */
public record ReceivePhaseSenderSkill(Player whoseTurn, Card messageCard, List<Player> receiveOrder,
                                      Player inFrontOfWhom) implements Fsm {
    @Override
    public ResolveResult resolve() {
        // TODO need complete this function
        return null;
    }
}

File: src/main/kotlin/phase/ReceivePhaseSenderSkill.kt
package com.fengsheng.phase

import com.fengsheng.Fsm
import com.fengsheng.Player
import com.fengsheng.ResolveResult
import com.fengsheng.card.Card

/**
 * 接收情报时，传出者的技能
 */
class ReceivePhaseSenderSkill(
    /**
     * 谁的回合
     */
    val whoseTurn: Player,
    /**
     * 情报传出者
     */
    val sender: Player,
    /**
     * 情报牌
     */
    val messageCard: Card,
    /**
     * 情报在谁面前
     */
    val inFrontOfWhom: Player,
    /**
     * 接收第三张黑色情报牌的顺序（也就是后续结算死亡的顺序）
     */
    val receiveOrder: ReceiveOrder = ReceiveOrder()
) : Fsm {
    override fun resolve(): ResolveResult {
        val result = whoseTurn.game!!.dealListeningSkill()
        return result ?: ResolveResult(
            ReceivePhaseReceiverSkill(whoseTurn, messageCard, receiveOrder, inFrontOfWhom),
            true
        )
    }

    override fun toString(): String {
        return "${whoseTurn}的回合，${inFrontOfWhom}成功接收情报"
    }
}